pkg/registry/pod/rest.go

```diff
@@ -81,9 +81,28 @@ func (podStrategy) ValidateUpdate(ctx api.Context, obj, old runtime.Object) fiel
 	return append(errorList, validation.ValidatePodUpdate(obj.(*api.Pod), old.(*api.Pod))...)
 }
 
-// CheckGracefulDelete allows a pod to be gracefully deleted.
+// CheckGracefulDelete allows a pod to be gracefully deleted. It updates the DeleteOptions to
+// reflect the desired grace value.
 func (podStrategy) CheckGracefulDelete(obj runtime.Object, options *api.DeleteOptions) bool {
-	return false
+	if options == nil {
+		return false
+	}
+	pod := obj.(*api.Pod)
+	// use the user provided value
+	if options.GracePeriodSeconds != nil {
+		period := int64(*options.GracePeriodSeconds)
+		// if the pod is not currently being deleted, or the value isn't set, or the user value is less than the stored value
+		if pod.Spec.TerminationGracePeriodSeconds == nil || pod.DeletionTimestamp == nil || period < *pod.Spec.TerminationGracePeriodSeconds {
+			pod.Spec.TerminationGracePeriodSeconds = &period
+		}
+	}
+	// if no grace period is requested, the pod is deleted immediately.
+	if pod.Spec.TerminationGracePeriodSeconds == nil {
```

Comment:
----- Original Message -----
> >  func (podStrategy) CheckGracefulDelete(obj runtime.Object, options
> >  *api.DeleteOptions) bool {
> > -	return false
> > +	if options == nil {
> > +		return false
> > +	}
> > +	pod := obj.(*api.Pod)
> > +	// use the user provided value
> > +	if options.GracePeriodSeconds != nil {
> > +		period := int64(*options.GracePeriodSeconds)
> > +		// if the pod is not currently being deleted, or the value isn't set, or
> > the user value is less than the stored value
> > +		if pod.Spec.TerminationGracePeriodSeconds == nil ||
> > pod.DeletionTimestamp == nil || period <
> > *pod.Spec.TerminationGracePeriodSeconds {
> > +			pod.Spec.TerminationGracePeriodSeconds = &period
> > +		}
> > +	}
> > +	// if no grace period is requested, the pod is deleted immediately.
> > +	if pod.Spec.TerminationGracePeriodSeconds == nil {
> 
> Should this even be possible given how defaulting works? I'd imagine
> explicitly setting the grace period to 0 would be how immediate deletion
> would be requested.

This would be the case of internal code that didn't go through defaulting, or someone deliberately mucking with the internal object.  This code path should never be hit though.